torch_geometric/nn/kge/transh.py

```diff
@@ -0,0 +1,165 @@
+
+        self.p_norm = p_norm
+        self.margin = margin
+
+        self.w_rel_emb = Embedding(num_relations, hidden_channels,
+                                   sparse=sparse)
+        self.d_rel_emb = Embedding(num_relations, hidden_channels,
+                                   sparse=sparse)
+
+        self.reset_parameters()
+
+    def reset_parameters(self):
+        bound = 6. / math.sqrt(self.hidden_channels)
+        torch.nn.init.uniform_(self.node_emb.weight, -bound, bound)
+        torch.nn.init.uniform_(self.w_rel_emb.weight, -bound, bound)
+        torch.nn.init.uniform_(self.d_rel_emb.weight, -bound, bound)
+        F.normalize(self.w_rel_emb.weight.data, p=self.p_norm, dim=-1,
+                    out=self.w_rel_emb.weight.data)
+        F.normalize(self.d_rel_emb.weight.data, p=self.p_norm, dim=-1,
+                    out=self.d_rel_emb.weight.data)
+
+    def forward(
+        self,
+        head_index: Tensor,
+        rel_type: Tensor,
+        tail_index: Tensor,
+    ) -> Tensor:
+
+        head = self.node_emb(head_index)
+        tail = self.node_emb(tail_index)
+
+        w_rel = F.normalize(self.w_rel_emb(rel_type), p=self.p_norm)
+        d_rel = self.d_rel_emb(rel_type)
+
+        proj_head = head - (w_rel * head).sum(dim=1).unsqueeze(dim=1) * w_rel
+        proj_tail = tail - (w_rel * tail).sum(dim=1).unsqueeze(dim=1) * w_rel
+
+        proj_head = F.normalize(proj_head, p=self.p_norm, dim=-1)
+        proj_tail = F.normalize(proj_tail, p=self.p_norm, dim=-1)
+
+        # Calculate *negative* TransH norm:
+        return -((
+            (proj_head + d_rel) - proj_tail).norm(p=self.p_norm, dim=-1))**2
+
+    def loss(
+        self,
+        head_index: Tensor,
+        rel_type: Tensor,
+        tail_index: Tensor,
+    ) -> Tensor:
+
+        pos_score = self(head_index, rel_type, tail_index)
+        if self.bernoulli:
+            neg_score = self(*self.random_sample_bernoulli(
+                head_index, rel_type, tail_index))
+        else:
+            neg_score = self(
+                *self.random_sample(head_index, rel_type, tail_index))
+
+        return F.margin_ranking_loss(
+            pos_score,
+            neg_score,
+            target=torch.ones_like(pos_score),
+            margin=self.margin,
+        )
+
+    def loader(
+        self,
+        head_index: Tensor,
+        rel_type: Tensor,
+        tail_index: Tensor,
+        **kwargs,
+    ) -> Tensor:
+        if self.bernoulli:
+            self.compute_corrupt_probs(head_index, rel_type, tail_index)
+        return KGTripletLoader(head_index, rel_type, tail_index, **kwargs)
+
+    def random_sample_bernoulli(
+            self, head_index: Tensor, rel_type: Tensor,
+            tail_index: Tensor) -> Tuple[Tensor, Tensor, Tensor]:
+
+        head_index = head_index.clone()
+        tail_index = tail_index.clone()
+
+        probs = torch.Tensor(
+            [self.head_corruption_probs[r.item()] for r in rel_type])
```

Comment:
Can we store `head_corruption_probs` as a Tensor instead and then just do `self.head_corruption_probs[rel_type]`?

Reply:
Fixed by @shruti-sridhar!